### ISSUE TITLE
Confused about usage of additionalProperties in a schema

### ISSUE THREAD

**Issue by falzm** [oapi-codegen/oapi-codegen#128]
Hi,

I'm a bit confused about the usage of `additionalProperties` [to create a dictionary/map as described in the Swagger doc](https://swagger.io/docs/specification/data-models/dictionaries/). Despite reading the README, I don't understand how a spec such as:

```yaml
components:
  schemas:
    MyResource:
      properties:
        created:
          format: date-time
          readOnly: true
          type: string
        description:
          type: string
        id:
          format: uuid
          readOnly: true
          type: string
        name:
          type: string
        tags:
          additionalProperties:
            type: string
          type: object
      type: object
```

results in:

```go
// MyResource defines model for MyResource.
type MyResource struct {
    Created     *time.Time              `json:"created,omitempty"`
    Description *string                 `json:"description,omitempty"`
    Id          *string                 `json:"id,omitempty"`
    Name        *string                 `json:"name,omitempty"`
    Tags        *MyResource_Tags        `json:"tags,omitempty"`
}

// MyResource_Tags defines model for MyResource.Tags.
type MyResource_Tags struct {
    AdditionalProperties map[string]string `json:"-"`
}
```

instead of

```go
// MyResource defines model for MyResource.
type MyResource struct {
    Created     *time.Time              `json:"created,omitempty"`
    Description *string                 `json:"description,omitempty"`
    Id          *string                 `json:"id,omitempty"`
    Name        *string                 `json:"name,omitempty"`
    Tags        map[string]string       `json:"tags,omitempty"`
}
```

Am I doing something wrong?

**Comment by deepmap-marcinr** [comment]
What you're trying to do is, unfortunately, very hard to express in Go, and currently, there isn't a way to do what you want.

To have tags be a map[string]string, I think we could special case a Schema that if it has no fields, and has AdditionalProperties set, we'd raise its internal schema up one level and produce that. It would be very nice, but a lot of special casing. MyResource_Tags behaves as you expect - it contains what you want, you just need to go through that AdditionalProperties field to get it.

**Comment by falzm** [comment]
@deepmap-marcinr thank you for the explanation. I'd very much like to see this special case implemented, but I'm not the one who has to maintain this project so eventually it's up to you, I won't insist (too much 😉). Maybe we can consider having this behavior configurable via a CLI option?

**Comment by veleek** [comment]
I've just run into this same scenario.  @deepmap-marcinr, you suggested this would be a lot of special casing.  Can you clarify why you think it's more complicated than the one scenario you described?  Without look at the code at all 😄 , I would imagine that you'd have something like

1. Resolve Type for Schema Element.
2. SchemaElement is a class with no fields
3. If flatten maps -> Return `map[string]<AdditionalPropertiesConstraints or interface{}>`
4. If !flatten maps -> Return `GeneratedClass with AdditionalProperties defined as ^`

**Comment by veleek** [comment]
In fiddling around with this a bit more, I'm confused how you would go about using this.  The golang JSON marshaler doesn't do anything special with the `AdditionalProperties` object, so if you pass in an object that expect additional properties like this it doesn't deserialize and you just lose the fields that aren't part of the class definition.

I understand that there may be reasons for having it generate code like this, I'm just trying to confirm what the correct way to use the classes generated by the existing code is.

**Comment by mromaszewicz** [comment]
I'm going to close this as `not planned` because we've tried really hard to generate Go code that maps 1-1 to the OpenAPI spec, and if we special case a behavior like "squashing" objects which contain only additional properties into a map, we're going to break existing specs, and also, if you add a different field after the fact, the structure of the code will change. The way it works today, unwieldy as it is, is predictable and doesn't require optimizing for developer intent.

I would love to have our own JSON serializer which handles additionalProperties as an embedded map, but I've also avoided reinventing wheels for the time being.
